File: app/src/main/java/kandroid/data/DockData.java
package kandroid.data;

import java.util.Date;

import kandroid.observer.POJO;
import kandroid.utils.Identifiable;

public class DockData extends Data<DockData.ApiNdock> implements Identifiable {

    public int getDockID() {
        return data.api_id;
    }

    public int getState() {
        return data.api_state;
    }

    public int getShipID() {
        return data.api_ship_id;
    }

    public Date getCompletionTime() {
        return new Date(data.api_complete_time);
    }

    @Override
    public void setData(ApiNdock data) {
        if (this.data != null && getState() == 1 && data.api_state == 0 && getShipID() != 0) {
            KCDatabase.getShips().get(getID()).repair();
        }
        super.setData(data);
    }

    @Override
    public int getID() {
        return data.api_id;
    }

    public static class ApiNdock extends POJO {
        public int api_member_id;
        public int api_id;
        public int api_state;
        public int api_ship_id;
        public int api_complete_time;
        public String api_complete_time_str;
        public int api_item1;
        public int api_item2;
        public int api_item3;
        public int api_item4;
    }
}

File: core/src/main/java/kandroid/data/DockData.kt
package kandroid.data

import com.google.gson.JsonElement
import kandroid.observer.kcsapi.api_req_nyukyo
import kandroid.utils.CatException
import kandroid.utils.collection.Identifiable
import kandroid.utils.json.get
import kandroid.utils.json.int
import kandroid.utils.json.long
import kandroid.utils.json.set
import java.util.*

/**
 * 入渠
 */
class DockData : JsonWrapper(), RequestDataListener, Identifiable {

    val dockId: Int get() = data["api_id"].int()
    /**
     * 入渠状态
     * -1=未解锁，0=空，1=入渠中
     */
    val state: Int get() = data["api_state"].int()
    val shipId: Int get() = data["api_ship_id"].int()
    val completionTime: Date get() = Date(data["api_complete_time"].long())

    override val id: Int get() = dockId

    override fun loadFromResponse(apiName: String, responseData: JsonElement) {
        val state = this.state
        val shipId = this.shipId
        super.loadFromResponse(apiName, responseData)
        val newState = this.state
        if (state == 1 && newState == 0 && shipId != 0) {
            KCDatabase.ships[shipId]?.repair()
        }
    }

    override fun loadFromRequest(apiName: String, requestData: MutableMap<String, String>) {
        when (apiName) {
            api_req_nyukyo.speedchange.name -> {
                if (state == 1 && shipId != 0) {
                    KCDatabase.ships[shipId]?.repair()
                    data["api_state"] = 0
                    data["api_ship_id"] = 0
                }
            }
            else -> CatException()
        }
    }
}